README.md
# i2cdevice

[![Build Status](https://travis-ci.com/pimoroni/i2cdevice-python.svg?branch=master)](https://travis-ci.com/pimoroni/i2cdevice-python)
[![Coverage Status](https://coveralls.io/repos/github/pimoroni/i2cdevice-python/badge.svg?branch=master)](https://coveralls.io/github/pimoroni/i2cdevice-python?branch=master)
[![PyPi Package](https://img.shields.io/pypi/v/i2cdevice.svg)](https://pypi.python.org/pypi/i2cdevice)
[![Python Versions](https://img.shields.io/pypi/pyversions/i2cdevice.svg)](https://pypi.python.org/pypi/i2cdevice)

i2cdevice is a Python framework aimed at dealing with common SMBus/i2c device interaction patterns.

This project aims to make group-up implementations of Python libraries for i2c devices easier, simpler and inherently self-documenting.

It does this by separating a detailed description of the hardware registers and how they should be manipulated into a structured definition language.

This project does not aim to help you make a public API for Python devices- that should be built on top of the fundamentals presented here.

# Using This Library

You should generally aim for a 1:1 representation of the hardware registers in the device you're implementing, even if you don't plan to use all the functionality. Having the full register set implemented allows for the easy addition of new features in future.

Check out the libraries listed below for real-world examples.

# Features

* Classes for describing devices, registers and individual bit fields within registers in a fashion which maps closely with the datasheet
* Value translation from real world numbers (such as `512ms`) to register values (such as `0b111`) and back again
* Read registers into a namedtuple of fields using `get`
* Write multiple register fields in a transaction using `set` with keyword arguments
* Support for treating multiple-bytes as a single value, or single register with multiple values

# Built With i2cdevice

* bme280 - https://github.com/pimoroni/bme280-python
* bmp280 - https://github.com/pimoroni/bmp280-python
* bh1745 - https://github.com/pimoroni/bh1745-python
* as7262 - https://github.com/pimoroni/as7262-python
* lsm303d - https://github.com/pimoroni/lsm303d-python
* ltr559 - https://github.com/pimoroni/ltr559-python
* ads1015 - https://github.com/pimoroni/ads1015-python

# Examples

The below example defines the `ALS_CONTROL` register on an ltr559, with register address `0x80`.

It has 3 fields; gain - which is mapped to real world values - and sw_reset/mode which are single bit flags.

```python
ALS_CONTROL = Register('ALS_CONTROL', 0x80, fields=(
    BitField('gain', 0b00011100, values_map={1: 0b000, 2: 0b001, 4: 0b011, 8:0b011, 48:0b110, 96:0b111}),
    BitField('sw_reset', 0b00000010),
    BitField('mode', 0b00000001)
))
```

A lookup table is not required for values, however, a function can be used to translate values from and to a format that the device understands.

The below example uses `i2cdevice._byte_swap` to change the endianness of two 16bit values before they are stored/retrieved.

```python
# This will address 0x88, 0x89, 0x8A and 0x8B as a continuous 32bit register
ALS_DATA = Register('ALS_DATA', 0x88, fields=(
    BitField('ch1', 0xFFFF0000, bitwidth=16, values_in=_byte_swap, values_out=_byte_swap),
    BitField('ch0', 0x0000FFFF, bitwidth=16, values_in=_byte_swap, values_out=_byte_swap)
), read_only=True, bitwidth=32)
```

A "Register" and its "BitField"s define a set of rules and logic for detailing with the hardware register which is interpreted by the device class. Registers are declared on a device using the `registers=()` keyword argument:

```python
I2C_ADDR = 0x23
ltr559 = Device(I2C_ADDR, bit_width=8, registers=(
	ALS_CONTROL,
	ALS_DATA
))
```

## Reading Registers

One configured a register's fields can be read into a namedtuple using the `get` method:

```python
register_values = ltr559.get('ALS_CONTROL')
gain = register_values.gain
sw_reset = register_values.sw_reset
mode = register_values.mode
```

## Writing Registers

The namedtuple returned from `get` is immutable and does not attempt to map values back to the hardware, in order to write one or more fields to a register you must use `set` with a keyword argument for each field:

```python
ltr559.set('ALS_CONTROL',
           gain=4,
           sw_reset=1)
```

This will read the register state from the device, update the bitfields accordingly and write the result back.


CHANGELOG.md
1.0.0
-----

* Enhancement: Repackage to pyproject.toml/hatchling
* Switch i2c_dev fallback from smbus to smbus2

0.0.7
-----

* BugFix: Prevent .get from reading multiple times (thanks @dkao)
* Enhancement: Better error output (thanks Kamil Klimek)

0.0.6
-----

* New API methods set and get

0.0.5
-----

* Bump to stable release

0.0.4
-----

* Bugfixes

0.0.3
-----

* Added License

0.0.2
-----

* Major Refactor

0.0.1
-----

* Initial Release


library/test.py
from i2cdevice import MockSMBus, Device, Register, BitField
from i2cdevice.adapter import Adapter, LookupAdapter


class x2Adapter(Adapter):
    def _decode(self, value):
        return value >> 1

    def _encode(self, value):
        return value << 1


bus = MockSMBus(1)

device = Device(0x00, bus, registers=(
    Register('status', 0x00, fields=(
        BitField('interrupt', 0b00000001, read_only=True),
        BitField('data_ready', 0b00000010),
        BitField('power', 0b10000000),
        BitField('monkey', 0b01111100, adapter=x2Adapter())
    )),
    Register('measurement_rate', 0x85, fields=(
        BitField('integration_time_ms', 0b00111000, adapter=LookupAdapter({100: 0b000, 50: 0b001, 200: 0b010, 400: 0b011, 150: 0b100, 250: 0b101, 300: 0b110, 350: 0b111})),
        BitField('repeat_rate_ms', 0b00000111, adapter=LookupAdapter({50: 0b000, 100: 0b001, 200: 0b010, 500: 0b011, 1000: 0b100, 2000: 0b101}))
    ))
))

device.set_field('status', 'power', False)
print(device.get_field('status', 'power'))
print(device.get_register('status'))
print(bus.regs[0])

device.set_field('status', 'power', True)
print(device.get_field('status', 'power'))
print(device.get_register('status'))
print(bus.regs[0])

print("Testing measurement_rate->integration_time_ms")
device.set_field('measurement_rate', 'integration_time_ms', 200)
print("{:08b}".format(bus.regs[0x85]))
assert device.get_field('measurement_rate', 'integration_time_ms') == 200

print("Object set", device.measurement_rate.set_integration_time_ms(400))
print("{:08b}".format(bus.regs[0x85]))
print("Object get", device.measurement_rate.get_integration_time_ms())

device.set_field('status', 'monkey', 1)
print("{:08b}".format(bus.regs[0]))
assert device.get_field('status', 'monkey') == 1

device.set_field('status', 'monkey', 2)
print("{:08b}".format(bus.regs[0]))
assert device.get_field('status', 'monkey') == 2

device.set_field('status', 'monkey', 3)
print("{:08b}".format(bus.regs[0]))
assert device.get_field('status', 'monkey') == 3


tests/test_utils.py
import pytest

from i2cdevice import _int_to_bytes, _leading_zeros, _mask_width, _trailing_zeros


def test_mask_width():
    assert _mask_width(0b111) == 3
    assert _mask_width(0b101) == 3
    assert _mask_width(0b0111) == 3
    assert _mask_width(0b1110) == 3


def test_leading_zeros():
    assert _leading_zeros(0b1) == 7
    assert _leading_zeros(0b10) == 6
    assert _leading_zeros(0b100) == 5
    assert _leading_zeros(0b100000000) == 8  # 9nth bit not counted by default


def test_trailing_zeros():
    assert _trailing_zeros(0b1) == 0
    assert _trailing_zeros(0b10) == 1
    assert _trailing_zeros(0b100) == 2
    assert _trailing_zeros(0b00000000) == 8  # Mask is all zeros


def test_int_to_bytes():
    assert _int_to_bytes(512, 2) == b'\x02\x00'
    assert _int_to_bytes(512, 2, endianness='little') == b'\x00\x02'
    assert _int_to_bytes(512, 2, endianness='big') == b'\x02\x00'

    with pytest.raises(TypeError):
        _int_to_bytes('', 2)


tests/test_device.py
import pytest

from i2cdevice import BitField, BitFlag, Device, MockSMBus, Register
from i2cdevice.adapter import U16ByteSwapAdapter


def test_register_locking():
    bus = MockSMBus(1)
    device = Device(0x00, i2c_dev=bus, registers=(
        Register('test', 0x00, fields=(
            BitField('test', 0xFF),
        )),
    ))

    device.test.set_test(77)
    device.lock_register('test')

    bus.regs[0] = 11
    assert device.test.get_test() == 77

    device.unlock_register('test')
    assert device.test.get_test() == 11


def test_adapters():
    bus = MockSMBus(1)
    device = Device(0x00, i2c_dev=bus, registers=(
        Register('adapter', 0x01, fields=(
            BitField('test', 0xFFFF, adapter=U16ByteSwapAdapter()),
        )),
    ))

    device.adapter.set_test(0xFF00)

    assert device.adapter.get_test() == 0xFF00

    assert bus.regs[0:2] == [0x00, 0xFF]


def test_address_select():
    bus = MockSMBus(1)
    device = Device([0x00, 0x01], i2c_dev=bus, registers=(
        Register('test', 0x00, fields=(
            BitField('test', 0xFF),
        )),
    ))

    assert device.get_addresses() == [0x00, 0x01]
    assert device.select_address(0x01) is True
    with pytest.raises(ValueError):
        device.select_address(0x02)

    assert device.next_address() == 0x00
    assert device.next_address() == 0x01


def test_get_set_field():
    bus = MockSMBus(1)
    device = Device([0x00, 0x01], i2c_dev=bus, registers=(
        Register('test', 0x00, fields=(
            BitField('test', 0xFF),
        )),
    ))

    device.set_field('test', 'test', 99)

    assert device.get_field('test', 'test') == 99

    assert bus.regs[0] == 99


def test_get_set_field_overflow():
    bus = MockSMBus(1)
    device = Device([0x00, 0x01], i2c_dev=bus, registers=(
        Register('test', 0x00, fields=(
            BitField('test', 0xFF),
        )),
    ))

    device.set_field('test', 'test', 9999999)

    assert device.get_field('test', 'test') == 127

    assert bus.regs[0] == 127


def test_bitflag():
    bus = MockSMBus(1)
    device = Device([0x00, 0x01], i2c_dev=bus, registers=(
        Register('test', 0x00, fields=(
            BitFlag('test', 6),  # Sixth bit from the right
        )),
    ))

    device.test.set_test(True)

    assert bus.regs[0] == 0b01000000

    device.test.set_test(False)

    assert bus.regs[0] == 0b00000000


def test_get_register():
    bus = MockSMBus(1)
    bus.regs[0:3] = [0xAA, 0xBB, 0xCC, 0xDD, 0xEE, 0xFF]
    device = Device([0x00, 0x01], i2c_dev=bus, registers=(
        Register('test24', 0x00, fields=(
            BitField('test', 0xFFF),
        ), bit_width=24),
        Register('test32', 0x00, fields=(
            BitField('test', 0xFFF),
        ), bit_width=32),
        Register('test48', 0x00, fields=(
            BitField('test', 0xFFF),
        ), bit_width=48),
    ))

    assert device.get_register('test24') == 0xAABBCC

    assert device.get_register('test32') == 0xAABBCCDD

    assert device.get_register('test48') == 0xAABBCCDDEEFF


def test_missing_regiser():
    bus = MockSMBus(1)
    device = Device([0x00, 0x01], i2c_dev=bus, registers=(
        Register('test', 0x00, fields=(
            BitFlag('test', 6),  # Sixth bit from the right
        )),
    ))

    with pytest.raises(KeyError):
        device.get_register('foo')


i2cdevice/__init__.py
from collections import namedtuple

__version__ = "1.0.0"


def _mask_width(value, bit_width=8):
    """Get the width of a bitwise mask

    ie: 0b000111 = 3
    """
    value >>= _trailing_zeros(value, bit_width)
    return value.bit_length()


def _leading_zeros(value, bit_width=8):
    """Count leading zeros on a binary number with a given bit_width

    ie: 0b0011 = 2

    Used for shifting around values after masking.
    """
    count = 0
    for _ in range(bit_width):
        if value & (1 << (bit_width - 1)):
            return count
        count += 1
        value <<= 1
    return count


def _trailing_zeros(value, bit_width=8):
    """Count trailing zeros on a binary number with a given bit_width

    ie: 0b11000 = 3

    Used for shifting around values after masking.
    """
    count = 0
    for _ in range(bit_width):
        if value & 1:
            return count
        count += 1
        value >>= 1
    return count


def _int_to_bytes(value, length, endianness='big'):
    try:
        return value.to_bytes(length, endianness)
    except AttributeError:
        output = bytearray()
        for x in range(length):
            offset = x * 8
            mask = 0xff << offset
            output.append((value & mask) >> offset)
        if endianness == 'big':
            output.reverse()
        return output


class MockSMBus:
    def __init__(self, i2c_bus, default_registers=None):
        self.regs = [0 for _ in range(255)]
        if default_registers is not None:
            for index in default_registers.keys():
                self.regs[index] = default_registers.get(index)

    def write_i2c_block_data(self, i2c_address, register, values):
        self.regs[register:register + len(values)] = values

    def read_i2c_block_data(self, i2c_address, register, length):
        return self.regs[register:register + length]


class _RegisterProxy(object):
    """Register Proxy

    This proxy catches lookups against non existent get_fieldname and set_fieldname methods
    and converts them into calls against the device's get_field and set_field methods with
    the appropriate options.

    This means device.register.set_field(value) and device.register.get_field(value) will work
    and also transparently update the underlying device without the register or field objects
    having to know anything about how data is written/read/stored.

    """
    def __init__(self, device, register):
        self.device = device
        self.register = register

    def __getattribute__(self, name):
        if name.startswith("get_"):
            name = name.replace("get_", "")
            return lambda: self.device.get_field(self.register.name, name)
        if name.startswith("set_"):
            name = name.replace("set_", "")
            return lambda value: self.device.set_field(self.register.name, name, value)
        return object.__getattribute__(self, name)

    def write(self):
        return self.device.write_register(self.register.name)

    def read(self):
        return self.device.read_register(self.register.name)

    def __enter__(self):
        self.device.read_register(self.register.name)
        self.device.lock_register(self.register.name)
        return self

    def __exit__(self, exception_type, exception_value, exception_traceback):
        self.device.unlock_register(self.register.name)


class Register():
    """Store information about an i2c register"""
    def __init__(self, name, address, fields=None, bit_width=8, read_only=False, volatile=True):
        self.name = name
        self.address = address
        self.bit_width = bit_width
        self.read_only = read_only
        self.volatile = volatile
        self.is_read = False
        self.fields = {}

        for field in fields:
            self.fields[field.name] = field

        self.namedtuple = namedtuple(self.name, sorted(self.fields))


class BitField():
    """Store information about a field or flag in an i2c register"""
    def __init__(self, name, mask, adapter=None, bit_width=8, read_only=False):
        self.name = name
        self.mask = mask
        self.adapter = adapter
        self.bit_width = bit_width
        self.read_only = read_only


class BitFlag(BitField):
    def __init__(self, name, bit, read_only=False):
        BitField.__init__(self, name, 1 << bit, adapter=None, bit_width=8, read_only=read_only)


class Device(object):
    def __init__(self, i2c_address, i2c_dev=None, bit_width=8, registers=None):
        self._bit_width = bit_width

        self.locked = {}
        self.registers = {}
        self.values = {}

        if isinstance(i2c_address, list):
            self._i2c_addresses = i2c_address
            self._i2c_address = i2c_address[0]
        else:
            self._i2c_addresses = [i2c_address]
            self._i2c_address = i2c_address

        self._i2c = i2c_dev

        if self._i2c is None:
            import smbus2
            self._i2c = smbus2.SMBus(1)

        for register in registers:
            self.locked[register.name] = False
            self.values[register.name] = 0
            self.registers[register.name] = register
            self.__dict__[register.name] = _RegisterProxy(self, register)

    def lock_register(self, name):
        self.locked[name] = True

    def unlock_register(self, name):
        self.locked[name] = False

    def read_register(self, name):
        register = self.registers[name]
        if register.volatile or not register.is_read:
            self.values[register.name] = self._i2c_read(register.address, register.bit_width)
            register.is_read = True
        return self.values[register.name]

    def write_register(self, name):
        register = self.registers[name]
        return self._i2c_write(register.address, self.values[register.name], register.bit_width)

    def get_addresses(self):
        return self._i2c_addresses

    def select_address(self, address):
        if address in self._i2c_addresses:
            self._i2c_address = address
            return True
        raise ValueError("Address {:02x} invalid!".format(address))

    def next_address(self):
        next_addr = self._i2c_addresses.index(self._i2c_address)
        next_addr += 1
        next_addr %= len(self._i2c_addresses)
        self._i2c_address = self._i2c_addresses[next_addr]
        return self._i2c_address

    def set(self, register, **kwargs):
        """Write one or more fields on a device register.

        Accepts multiple keyword arguments, one for each field to write.

        :param register: Name of register to write.

        """
        self.read_register(register)
        self.lock_register(register)
        for field in kwargs.keys():
            value = kwargs.get(field)
            self.set_field(register, field, value)
        self.write_register(register)
        self.unlock_register(register)

    def get(self, register):
        """Get a namedtuple containing register fields.

        :param register: Name of register to retrieve

        """
        result = {}
        self.read_register(register)
        self.lock_register(register)
        for field in self.registers[register].fields:
            result[field] = self.get_field(register, field)
        self.unlock_register(register)
        return self.registers[register].namedtuple(**result)

    def get_field(self, register, field):
        register = self.registers[register]
        field = register.fields[field]

        if not self.locked[register.name]:
            self.read_register(register.name)

        value = self.values[register.name]

        value = (value & field.mask) >> _trailing_zeros(field.mask, register.bit_width)

        if field.adapter is not None:
            try:
                value = field.adapter._decode(value)
            except ValueError as value_error:
                raise ValueError("{}: {}".format(field.name, str(value_error)))

        return value

    def set_field(self, register, field, value):
        register = self.registers[register]
        field = register.fields[field]
        shift = _trailing_zeros(field.mask, register.bit_width)

        if field.adapter is not None:
            value = field.adapter._encode(value)

        if not self.locked[register.name]:
            self.read_register(register.name)

        reg_value = self.values[register.name]

        reg_value &= ~field.mask
        reg_value |= (value << shift) & field.mask

        self.values[register.name] = reg_value

        if not self.locked[register.name]:
            self.write_register(register.name)

    def get_register(self, register):
        register = self.registers[register]
        return self._i2c_read(register.address, register.bit_width)

    def _i2c_write(self, register, value, bit_width):
        values = _int_to_bytes(value, bit_width // self._bit_width, 'big')
        values = list(values)
        self._i2c.write_i2c_block_data(self._i2c_address, register, values)

    def _i2c_read(self, register, bit_width):
        value = 0
        for x in self._i2c.read_i2c_block_data(self._i2c_address, register, bit_width // self._bit_width):
            value <<= 8
            value |= x
        return value


library/i2cdevice/adapter.py
class Adapter:
    """
    Must implement `_decode()` and `_encode()`.
    """
    def _decode(self, value):
        raise NotImplementedError

    def _encode(self, value):
        raise NotImplementedError


class LookupAdapter(Adapter):
    """Adaptor with a dictionary of values.

    :param lookup_table: A dictionary of one or more key/value pairs where the key is the human-readable value and the value is the bitwise register value

    """
    def __init__(self, lookup_table, snap=True):
        self.lookup_table = lookup_table
        self.snap = snap

    def _decode(self, value):
        for k, v in self.lookup_table.items():
            if v == value:
                return k
        raise ValueError("{} not in lookup table".format(value))

    def _encode(self, value):
        if self.snap and type(value) in [int, float]:
            value = min(list(self.lookup_table.keys()), key=lambda x: abs(x - value))
        return self.lookup_table[value]


class U16ByteSwapAdapter(Adapter):
    """Adaptor to swap the bytes in a 16bit integer."""
    def _byteswap(self, value):
        return (value >> 8) | ((value & 0xFF) << 8)

    def _decode(self, value):
        return self._byteswap(value)

    def _encode(self, value):
        return self._byteswap(value)


examples/ltr559.py
import sys
import time

sys.path.insert(0, "../")
from i2cdevice import BitField, Device, MockSMBus, Register  # noqa: E402
from i2cdevice.adapter import Adapter, LookupAdapter, U16ByteSwapAdapter  # noqa: E402

I2C_ADDR = 0x23


class Bit12Adapter(Adapter):
    def _encode(self, value):
        """
        Convert the 16-bit output into the correct format for reading:

            0bLLLLLLLLXXXXHHHH -> 0bHHHHLLLLLLLL
        """
        return ((value & 0xFF)) << 8 | ((value & 0xF00) >> 8)

    def _decode(self, value):
        """
        Convert the 12-bit input into the correct format for the registers,
        the low byte followed by 4 empty bits and the high nibble:

            0bHHHHLLLLLLLL -> 0bLLLLLLLLXXXXHHHH
        """
        return ((value & 0xFF00) >> 8) | ((value & 0x000F) << 8)


ltr559 = Device(I2C_ADDR, i2c_dev=MockSMBus(0, default_registers={0x86: 0x92}), bit_width=8, registers=(

    Register('ALS_CONTROL', 0x80, fields=(
        BitField('gain', 0b00011100, adapter=LookupAdapter({1: 0b000, 2: 0b001, 4: 0b011, 8: 0b011, 48: 0b110, 96: 0b111})),
        BitField('sw_reset', 0b00000010),
        BitField('mode', 0b00000001)
    )),

    Register('PS_CONTROL', 0x81, fields=(
        BitField('saturation_indicator_enable', 0b00100000),
        BitField('active', 0b00000011, adapter=LookupAdapter({False: 0b00, True: 0b11}))
    )),

    Register('PS_LED', 0x82, fields=(
        BitField('pulse_freq_khz', 0b11100000, adapter=LookupAdapter({30: 0b000, 40: 0b001, 50: 0b010, 60: 0b011, 70: 0b100, 80: 0b101, 90: 0b110, 100: 0b111})),
        BitField('duty_cycle', 0b00011000, adapter=LookupAdapter({0.25: 0b00, 0.5: 0b01, 0.75: 0b10, 1.0: 0b11})),
        BitField('current_ma', 0b00000111, adapter=LookupAdapter({5: 0b000, 10: 0b001, 20: 0b010, 50: 0b011, 100: 0b100}))
    )),

    Register('PS_N_PULSES', 0x83, fields=(
        BitField('count', 0b00001111),
    )),

    Register('PS_MEAS_RATE', 0x84, fields=(
        BitField('rate_ms', 0b00001111, adapter=LookupAdapter({10: 0b1000, 50: 0b0000, 70: 0b0001, 100: 0b0010, 200: 0b0011, 500: 0b0100, 1000: 0b0101, 2000: 0b0110})),
    )),

    Register('ALS_MEAS_RATE', 0x85, fields=(
        BitField('integration_time_ms', 0b00111000, adapter=LookupAdapter({100: 0b000, 50: 0b001, 200: 0b010, 400: 0b011, 150: 0b100, 250: 0b101, 300: 0b110, 350: 0b111})),
        BitField('repeat_rate_ms', 0b00000111, adapter=LookupAdapter({50: 0b000, 100: 0b001, 200: 0b010, 500: 0b011, 1000: 0b100, 2000: 0b101}))
    )),

    Register('PART_ID', 0x86, fields=(
        BitField('part_number', 0b11110000),  # Should be 0x09H
        BitField('revision', 0b00001111)  # Should be 0x02H
    ), read_only=True, volatile=False),

    Register('MANUFACTURER_ID', 0x87, fields=(
        BitField('manufacturer_id', 0b11111111),  # Should be 0x05H
    ), read_only=True),

    # This will address 0x88, 0x89, 0x8A and 0x8B as a continuous 32bit register
    Register('ALS_DATA', 0x88, fields=(
        BitField('ch1', 0xFFFF0000, bit_width=16, adapter=U16ByteSwapAdapter()),
        BitField('ch0', 0x0000FFFF, bit_width=16, adapter=U16ByteSwapAdapter())
    ), read_only=True, bit_width=32),

    Register('ALS_PS_STATUS', 0x8C, fields=(
        BitField('als_data_valid', 0b10000000),
        BitField('als_gain', 0b01110000, adapter=LookupAdapter({1: 0b000, 2: 0b001, 4: 0b010, 8: 0b011, 48: 0b110, 96: 0b111})),
        BitField('als_interrupt', 0b00001000),  # True = Interrupt is active
        BitField('als_data', 0b00000100),  # True = New data available
        BitField('ps_interrupt', 0b00000010),  # True = Interrupt is active
        BitField('ps_data', 0b00000001)  # True = New data available
    ), read_only=True),

    # The PS data is actually an 11bit value but since B3 is reserved it'll (probably) read as 0
    # We could mask the result if necessary
    Register('PS_DATA', 0x8D, fields=(
        BitField('ch0', 0xFF0F, adapter=Bit12Adapter()),
        BitField('saturation', 0x0080)
    ), bit_width=16, read_only=True),

    # INTERRUPT allows the interrupt pin and function behaviour to be configured.
    Register('INTERRUPT', 0x8F, fields=(
        BitField('polarity', 0b00000100),
        BitField('mode', 0b00000011, adapter=LookupAdapter({'off': 0b00, 'ps': 0b01, 'als': 0b10, 'als+ps': 0b11}))
    )),

    Register('PS_THRESHOLD', 0x90, fields=(
        BitField('upper', 0xFF0F0000, adapter=Bit12Adapter(), bit_width=16),
        BitField('lower', 0x0000FF0F, adapter=Bit12Adapter(), bit_width=16)
    ), bit_width=32),

    # PS_OFFSET defines the measurement offset value to correct for proximity
    # offsets caused by device variations, crosstalk and other environmental factors.
    Register('PS_OFFSET', 0x94, fields=(
        BitField('offset', 0x03FF),  # Last two bits of 0x94, full 8 bits of 0x95
    ), bit_width=16),

    # Defines the upper and lower limits of the "ALS" reading.
    # An interrupt is triggered if values fall outside of this range.
    # See also INTERRUPT_PERSIST.
    Register('ALS_THRESHOLD', 0x97, fields=(
        BitField('upper', 0xFFFF0000, adapter=U16ByteSwapAdapter(), bit_width=16),
        BitField('lower', 0x0000FFFF, adapter=U16ByteSwapAdapter(), bit_width=16)
    ), bit_width=32),

    # This register controls how many values must fall outside of the range defined
    # by upper and lower threshold limits before the interrupt is asserted.

    # In the case of both "PS" and "ALS", a 0 value indicates that every value outside
    # the threshold range should be counted.
    # Values therein map to n+1 , ie: 0b0001 requires two consecutive values.
    Register('INTERRUPT_PERSIST', 0x9E, fields=(
        BitField('PS', 0xF0),
        BitField('ALS', 0x0F)
    ))

))


if __name__ == "__main__":
    part_id = ltr559.get('PART_ID')

    assert part_id.part_number == 0x09
    assert part_id.revision == 0x02

    print("""
Found LTR-559.
Part ID: 0x{:02x}
Revision: 0x{:02x}
    """.format(
        part_id.part_number,
        part_id.revision
    ))

    print("""
Soft Reset
    """)
    ltr559.set('ALS_CONTROL', sw_reset=1)
    ltr559.set('ALS_CONTROL', sw_reset=0)
    try:
        while True:
            status = ltr559.get('ALS_CONTROL').sw_reset
            print("Status: {}".format(status))
            if status == 0:
                break
            time.sleep(1.0)
    except KeyboardInterrupt:
        pass

    print("Setting light sensor threshold")

    # The `set` method can handle writes to multiple register fields
    # specify each field value as a keyword argument
    ltr559.set('ALS_THRESHOLD',
                lower=0x0001,
                upper=0xFFEE)

    print("{:08x}".format(ltr559.values['ALS_THRESHOLD']))

    # The `get` method returns register values as an immutable
    # namedtuple, and fields will be available as properties
    # You must use `set` to write a register.
    als_threshold = ltr559.get('ALS_THRESHOLD')
    assert als_threshold.lower == 0x0001
    assert als_threshold.upper == 0xFFEE

    print("Setting PS threshold")

    ltr559.set('PS_THRESHOLD',
               lower=0,
               upper=500)

    print("{:08x}".format(ltr559.values['PS_THRESHOLD']))

    ps_threshold = ltr559.get('PS_THRESHOLD')
    assert ps_threshold.lower == 0
    assert ps_threshold.upper == 500

    print("Setting integration time and repeat rate")
    ltr559.set('PS_MEAS_RATE', rate_ms=100)
    ltr559.set('ALS_MEAS_RATE',
                integration_time_ms=50,
                repeat_rate_ms=50)

    als_meas_rate = ltr559.get('ALS_MEAS_RATE')
    assert als_meas_rate.integration_time_ms == 50
    assert als_meas_rate.repeat_rate_ms == 50

    print("""
Activating sensor
    """)

    ltr559.set('INTERRUPT', mode='als+ps')
    ltr559.set('PS_CONTROL',
               active=True,
               saturation_indicator_enable=1)

    ltr559.set('PS_LED',
               current_ma=50,
               duty_cycle=1.0,
               pulse_freq_khz=30)

    ltr559.set('PS_N_PULSES', count=1)

    ltr559.set('ALS_CONTROL',
               mode=1,
               gain=4)

    als_control = ltr559.get('ALS_CONTROL')
    assert als_control.mode == 1
    assert als_control.gain == 4

    ltr559.set('PS_OFFSET', offset=69)

    als0 = 0
    als1 = 0
    ps0 = 0
    lux = 0

    ch0_c = (17743, 42785, 5926, 0)
    ch1_c = (-11059, 19548, -1185, 0)

    try:
        while True:
            als_ps_status = ltr559.get('ALS_PS_STATUS')
            ps_int = als_ps_status.ps_interrupt or als_ps_status.ps_data
            als_int = als_ps_status.als_interrupt or als_ps_status.als_data

            if ps_int:
                ps0 = ltr559.get('PS_DATA').ch0

            if als_int:
                als_data = ltr559.get('ALS_DATA')
                als0 = als_data.ch0
                als1 = als_data.ch1

                ratio = 1000
                if als0 + als0 > 0:
                    ratio = (als0 * 1000) / (als1 + als0)

                ch_idx = 3
                if ratio < 450:
                    ch_idx = 0
                elif ratio < 640:
                    ch_idx = 1
                elif ratio < 850:
                    ch_idx = 2

                lux = ((als0 * ch0_c[ch_idx]) - (als1 * ch1_c[ch_idx])) / 10000

            print("Lux: {:06.2f}, Light CH0: {:04d}, Light CH1: {:04d}, Proximity: {:04d}  New Data LP: 0b{:01d}{:01d}".format(lux, als0, als1, ps0, als_int, ps_int))
            time.sleep(0.05)
    except KeyboardInterrupt:
        pass
